include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,205 @@
 
 #include <irr/builtin/material_compiler/glsl/common.glsl>
 
-void instr_eval_execute(in instr_t instr, in vec3 L)
+void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _microfacet)
 {
-	uint op = instr_getOpcode(instr);
+	const uint op = instr_getOpcode(instr);
+	const bool is_bxdf = op_isBXDF(op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+	const bool positiveCosFactor = cosFactor > FLT_MIN;
 
-	//speculative execution
-	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
-	params_t params = instr_getParameters(instr, bsdf_data);
-	float bxdf_eval_scalar_part;
-	uint ndf = instr_getNDF(instr);
-	float a = params_getAlpha(params);
-	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
-	float ay = params_getAlphaV(params);
-	float ay2 = ay*ay;
-#endif
+	uvec3 regs = instr_decodeRegisters(instr);
+	mat2x3 ior;
+	mat2x3 ior2;
+	params_t params;
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bsdf_data_t bsdf_data;
+
+	if (!is_bxdf || positiveCosFactor)
+	{
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
+		microfacet = _microfacet;
+	}
 
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+	bxdf_eval_t bxdf_eval = bxdf_eval_t(0.0);
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
+	if (is_bxdf && positiveCosFactor)
 	{
+		const float ior_scalar = colorToScalar(ior[0]);
+		float bxdf_eval_scalar_part;
+		uint ndf = instr_getNDF(instr);
+		float a = params_getAlpha(params);
+		float a2 = a*a;
+		float ay = params_getAlphaV(params);
+		float ay2 = ay*ay;
+		const vec3 albedo = params_getReflectance(params);
+
+		bool is_valid = true;
+		bool refraction = false;
+#ifndef NO_BSDF
+		//here actually using stronger check for BSDF because it's probably worth it
+		if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+		{
+			is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, ior_scalar, 1.0/ior_scalar);
+			refraction = true;
+		}
+#endif
+
+		if (is_valid)
+		{
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+			if (op_isDiffuse(op))
+			{
+				bxdf_eval = albedo * irr_glsl_oren_nayar_cos_eval(s, currInteraction.isotropic, a2);
+			} else
+#endif
+#if defined(OP_CONDUCTOR) || defined(OP_DIELECTRIC)
+			if (op_hasSpecular(op))
```

Comment:
move this out to contain the `is_valid` check (or move the `is_valid` and `refraction` stuff into the `op_hasSpecular` if case, depending which way you look at it)... because right now you're doing a specular refraction feasibility check on the DIFFTRANS material